actors/builtin/market/market_test.go

```diff
@@ -693,6 +695,363 @@ func TestPublishStorageDealsFailures(t *testing.T) {
+		newEpoch := endEpoch - 1
+		rt.SetEpoch(newEpoch)
+
+		// terminating all three deals ONLY terminates deal1 and deal2 because deal3 has expired
+		actor.terminateDeals(rt, provider, dealId1, dealId2, dealId3)
+		actor.assertDealsTerminated(rt, newEpoch, dealId1, dealId2)
+		actor.assertDeaslNotTerminated(rt, dealId3)
+
+	})
+
+	t.Run("do not terminate deal if end epoch is equal to or less than current epoch", func(t *testing.T) {
+		rt, actor := basicMarketSetup(t, marketActor, owner, provider, worker, client)
+		rt.SetEpoch(currentEpoch)
+
+		// deal1 has endepoch equal to current epoch when terminate is called
+		dealId1 := actor.generateAndPublishDeal(rt, client, mAddrs, startEpoch, endEpoch)
+		actor.activateDeals(rt, sectorExpiry, provider, currentEpoch, dealId1)
+		rt.SetEpoch(endEpoch)
+		actor.terminateDeals(rt, provider, dealId1)
+		actor.assertDeaslNotTerminated(rt, dealId1)
+
+		// deal2 has end epoch less than current epoch when terminate is called
+		rt.SetEpoch(currentEpoch)
+		dealId2 := actor.generateAndPublishDeal(rt, client, mAddrs, startEpoch+1, endEpoch)
+		actor.activateDeals(rt, sectorExpiry, provider, currentEpoch, dealId2)
+		rt.SetEpoch(endEpoch + 1)
+		actor.terminateDeals(rt, provider, dealId2)
+		actor.assertDeaslNotTerminated(rt, dealId2)
+	})
+
+	t.Run("fail when caller is not a StorageMinerActor", func(t *testing.T) {
+		rt, actor := basicMarketSetup(t, marketActor, owner, provider, worker, client)
+		rt.ExpectValidateCallerType(builtin.StorageMinerActorCodeID)
+		rt.SetCaller(provider, builtin.AccountActorCodeID)
+		rt.ExpectAbort(exitcode.ErrForbidden, func() {
+			rt.Call(actor.OnMinerSectorsTerminate, &market.OnMinerSectorsTerminateParams{})
+		})
+
+		rt.Verify()
+	})
+
+	t.Run("fail when caller is not the provider of the deal", func(t *testing.T) {
+		rt, actor := basicMarketSetup(t, marketActor, owner, provider, worker, client)
+		rt.SetEpoch(currentEpoch)
+
+		dealId := actor.generateAndPublishDeal(rt, client, mAddrs, startEpoch, endEpoch)
+		actor.activateDeals(rt, sectorExpiry, provider, currentEpoch, dealId)
+
+		params := mkTerminateDealParams(dealId)
+
+		provider2 := tutil.NewIDAddr(t, 501)
+		rt.ExpectValidateCallerType(builtin.StorageMinerActorCodeID)
+		rt.SetCaller(provider2, builtin.StorageMinerActorCodeID)
+		rt.ExpectAssertionFailure("caller is not the provider of the deal", func() {
+			rt.Call(actor.OnMinerSectorsTerminate, params)
+		})
+
+		rt.Verify()
+	})
+
+	t.Run("fail when deal has been published but not activated", func(t *testing.T) {
+		rt, actor := basicMarketSetup(t, marketActor, owner, provider, worker, client)
+		rt.SetEpoch(currentEpoch)
+
+		dealId := actor.generateAndPublishDeal(rt, client, mAddrs, startEpoch, endEpoch)
+
+		params := mkTerminateDealParams(dealId)
+		rt.ExpectValidateCallerType(builtin.StorageMinerActorCodeID)
+		rt.SetCaller(provider, builtin.StorageMinerActorCodeID)
+		rt.ExpectAbort(exitcode.ErrIllegalState, func() {
```

Comment:
Without looking at what's going on inside the market actor this seems like it should fail for IllegalArgument.  I think this is failing on the `!found` case of states.Get which would be an easy change to make if my intuition is right.

Reply:
I agree. Have made this change.